app/src/hooks/privy/usePrivyFarcaster.ts

```diff
@@ -0,0 +1,61 @@
+import { useLinkAccount, usePrivy } from "@privy-io/react-auth"
+import { useRef } from "react"
+import { toast } from "sonner"
+
+import { syncPrivyUser } from "@/db/privy"
+
+import { useUser } from "../db/useUser"
+import { useHandlePrivyErrors } from "../useHandlePrivyErrors"
+
+export const usePrivyFarcaster = (userId: string) => {
+  const isLinking = useRef(false)
+
+  const onError = useHandlePrivyErrors()
+  const { invalidate: invalidateUser } = useUser({ id: userId, enabled: false })
+  const { user: privyUser, unlinkFarcaster } = usePrivy()
+
+  const { linkFarcaster } = useLinkAccount({
+    onSuccess: async ({ user: updatedPrivyUser, linkMethod }) => {
+      if (linkMethod === "farcaster" && isLinking.current) {
+        toast.promise(
+          syncPrivyUser(updatedPrivyUser)
+            .then(() => invalidateUser())
+            .then(() => (isLinking.current = false)),
+          {
+            loading: "Linking farcaster...",
+            success: "Farcaster linked successfully",
+            error: "Failed to link farcaster",
+          },
+        )
+      }
+    },
+    onError,
+  })
+
+  const handleUnlinkFarcaster = () => {
+    if (privyUser?.farcaster?.fid) {
+      toast.promise(unlinkFarcaster(Number(privyUser.farcaster.fid)), {
```

Comment:
We shouldn't be calling critical actions within the toast. A couple reasons:

1. We shouldnt tie core platform functionality to UI rendering mechanism which can fail for different reasons
2. Error handling is abstracted and can fail